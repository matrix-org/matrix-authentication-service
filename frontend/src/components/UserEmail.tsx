// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  Button,
  Control,
  Field,
  Label,
  Message,
  Root as Form,
  Submit,
} from "@vector-im/compound-web";
import { atom, useAtom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition } from "react";

import { FragmentType, graphql, useFragment } from "../gql";

import Block from "./Block";
import DateTime from "./DateTime";
import Typography from "./Typography";

// This component shows a single user email address, with controls to verify it,
// resend the verification email, remove it, and set it as the primary email address.

const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserEmail_email on UserEmail {
    id
    email
    createdAt
    confirmedAt
  }
`);

const VERIFY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation VerifyEmail($id: ID!, $code: String!) {
    verifyEmail(input: { userEmailId: $id, code: $code }) {
      status

      user {
        id
        primaryEmail {
          id
        }
      }

      email {
        id
        ...UserEmail_email
      }
    }
  }
`);

const RESEND_VERIFICATION_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation ResendVerificationEmail($id: ID!) {
    sendVerificationEmail(input: { userEmailId: $id }) {
      status

      user {
        id
        primaryEmail {
          id
        }
      }

      email {
        id
        ...UserEmail_email
      }
    }
  }
`);

const REMOVE_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation RemoveEmail($id: ID!) {
    removeEmail(input: { userEmailId: $id }) {
      status

      user {
        id
      }
    }
  }
`);

const SET_PRIMARY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation SetPrimaryEmail($id: ID!) {
    setPrimaryEmail(input: { userEmailId: $id }) {
      status
      user {
        id
        primaryEmail {
          id
        }
      }
    }
  }
`);

const verifyEmailFamily = atomFamily((id: string) => {
  const verifyEmail = atomWithMutation(VERIFY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const verifyEmailAtom = atom(
    (get) => get(verifyEmail),
    (get, set, code: string) => set(verifyEmail, { id, code })
  );

  return verifyEmailAtom;
});

const resendVerificationEmailFamily = atomFamily((id: string) => {
  const resendVerificationEmail = atomWithMutation(
    RESEND_VERIFICATION_EMAIL_MUTATION
  );

  // A proxy atom which pre-sets the id variable in the mutation
  const resendVerificationEmailAtom = atom(
    (get) => get(resendVerificationEmail),
    (get, set) => set(resendVerificationEmail, { id })
  );

  return resendVerificationEmailAtom;
});

const removeEmailFamily = atomFamily((id: string) => {
  const removeEmail = atomWithMutation(REMOVE_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const removeEmailAtom = atom(
    (get) => get(removeEmail),
    (get, set) => set(removeEmail, { id })
  );

  return removeEmailAtom;
});

const setPrimaryEmailFamily = atomFamily((id: string) => {
  const setPrimaryEmail = atomWithMutation(SET_PRIMARY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const setPrimaryEmailAtom = atom(
    (get) => get(setPrimaryEmail),
    (get, set) => set(setPrimaryEmail, { id })
  );

  return setPrimaryEmailAtom;
});

const UserEmail: React.FC<{
  email: FragmentType<typeof FRAGMENT>;
  onRemove?: () => void;
  onSetPrimary?: () => void;
  isPrimary?: boolean;
  highlight?: boolean;
}> = ({ email, isPrimary, highlight, onSetPrimary, onRemove }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment(FRAGMENT, email);
  const [verifyEmailResult, verifyEmail] = useAtom(verifyEmailFamily(data.id));
  const [resendVerificationEmailResult, resendVerificationEmail] = useAtom(
    resendVerificationEmailFamily(data.id)
  );
  const setPrimaryEmail = useSetAtom(setPrimaryEmailFamily(data.id));
  const removeEmail = useSetAtom(removeEmailFamily(data.id));
  // TODO: compound doesn't forward the refs properly
  // const fieldRef = useRef<HTMLInputElement>(null);

  const onFormSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();
    const formData = new FormData(e.currentTarget);
    const code = formData.get("code") as string;
    startTransition(() => {
      verifyEmail(code).then((result) => {
        // Clear the form
        e.currentTarget?.reset();

        if (result.data?.verifyEmail.status === "VERIFIED") {
          // Call the onSetPrimary callback if provided
          // XXX: do we need a dedicated onVerify callback?
          onSetPrimary?.();
        }
      });
    });
  };

  const onResendClick = (): void => {
    startTransition(() => {
      resendVerificationEmail().then(() => {
        // TODO: fieldRef.current?.focus();
      });
    });
  };

  const onRemoveClick = (): void => {
    startTransition(() => {
      removeEmail().then(() => {
        // Call the onRemove callback if provided
        onRemove?.();
      });
    });
  };

  const onSetPrimaryClick = (): void => {
    startTransition(() => {
      setPrimaryEmail().then(() => {
        // Call the onSetPrimary callback if provided
        onSetPrimary?.();
      });
    });
  };

  const emailSent =
    resendVerificationEmailResult.data?.sendVerificationEmail.status === "SENT";
  const invalidCode =
    verifyEmailResult.data?.verifyEmail.status === "INVALID_CODE";

  return (
    <Block
      highlight={highlight}
      className="grid grid-col-1 gap-2 pb-4 border-b-2 border-b-grey-200"
    >
      {isPrimary && (
        <Typography variant="body" bold>
          Primary
        </Typography>
      )}
      <Typography variant="caption" bold className="flex-1">
        {data.email}
      </Typography>
      {data.confirmedAt ? (
        <Typography variant="micro">
          Verified <DateTime datetime={data.confirmedAt} />
        </Typography>
      ) : (
        <Form onSubmit={onFormSubmit} className="grid grid-cols-2 gap-2">
          <Field name="code" className="col-span-2">
            <Label>Code</Label>
            <Control
              // ref={fieldRef}
              placeholder="xxxxxx"
              type="text"
              inputMode="numeric"
            />
          </Field>
          {invalidCode && (
            <Message className="col-span-2 text-alert font-bold">
              Invalid code
            </Message>
          )}
          <Submit size="sm" type="submit" disabled={pending}>
            Submit
          </Submit>
          <Button
            size="sm"
            kind="secondary"
            disabled={pending || emailSent}
            onClick={onResendClick}
          >
            {emailSent ? "Sent!" : "Resend"}
          </Button>
        </Form>
      )}
      {!isPrimary && (
        <div className="flex justify-between items-center">
          {/* The primary email can only be set if the email was verified */}
          {data.confirmedAt ? (
            <Button size="sm" disabled={pending} onClick={onSetPrimaryClick}>
              Set primary
            </Button>
          ) : (
            <div />
          )}
          <Button
            kind="destructive"
            size="sm"
            disabled={pending}
            onClick={onRemoveClick}
          >
            Remove
          </Button>
        </div>
      )}
    </Block>
  );
};

export default UserEmail;
