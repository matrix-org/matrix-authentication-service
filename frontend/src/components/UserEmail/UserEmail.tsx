// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import IconDelete from "@vector-im/compound-design-tokens/icons/delete.svg?react";
import { Form, IconButton, Text, Tooltip } from "@vector-im/compound-web";
import { atom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition, ComponentProps, ReactNode } from "react";
import { Translation, useTranslation } from "react-i18next";

import { FragmentType, graphql, useFragment } from "../../gql";
import { Link } from "../../routing";
import ConfirmationModal from "../ConfirmationModal/ConfirmationModal";

import styles from "./UserEmail.module.css";

// This component shows a single user email address, with controls to verify it,
// resend the verification email, remove it, and set it as the primary email address.

export const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserEmail_email on UserEmail {
    id
    email
    confirmedAt
  }
`);

const REMOVE_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation RemoveEmail($id: ID!) {
    removeEmail(input: { userEmailId: $id }) {
      status

      user {
        id
      }
    }
  }
`);

const SET_PRIMARY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation SetPrimaryEmail($id: ID!) {
    setPrimaryEmail(input: { userEmailId: $id }) {
      status
      user {
        id
        primaryEmail {
          id
        }
      }
    }
  }
`);

const removeEmailFamily = atomFamily((id: string) => {
  const removeEmail = atomWithMutation(REMOVE_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const removeEmailAtom = atom(
    (get) => get(removeEmail),
    (_get, set) => set(removeEmail, { id }),
  );

  return removeEmailAtom;
});

const setPrimaryEmailFamily = atomFamily((id: string) => {
  const setPrimaryEmail = atomWithMutation(SET_PRIMARY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const setPrimaryEmailAtom = atom(
    (get) => get(setPrimaryEmail),
    (_get, set) => set(setPrimaryEmail, { id }),
  );

  return setPrimaryEmailAtom;
});

const DeleteButton: React.FC<{ disabled?: boolean; onClick?: () => void }> = ({
  disabled,
  onClick,
}) => (
  <Translation>
    {(t): ReactNode => (
      <Tooltip label={t("frontend.user_email.delete_button_title")}>
        <IconButton
          type="button"
          disabled={disabled}
          className="m-2"
          onClick={onClick}
          size="var(--cpd-space-8x)"
        >
          <IconDelete className={styles.userEmailDeleteIcon} />
        </IconButton>
      </Tooltip>
    )}
  </Translation>
);

const DeleteButtonWithConfirmation: React.FC<
  ComponentProps<typeof DeleteButton>
> = ({ onClick, ...rest }) => {
  const { t } = useTranslation();
  const onConfirm = (): void => {
    onClick?.();
  };

  // NOOP function, otherwise we dont render a cancel button
  const onDeny = (): void => {};

  return (
    <>
      <ConfirmationModal
        trigger={<DeleteButton {...rest} />}
        onDeny={onDeny}
        onConfirm={onConfirm}
      >
        <Text>
          {t("frontend.user_email.delete_button_confirmation_modal.body")}
        </Text>
      </ConfirmationModal>
    </>
  );
};

const UserEmail: React.FC<{
  email: FragmentType<typeof FRAGMENT>;
  onRemove?: () => void;
  onSetPrimary?: () => void;
  isPrimary?: boolean;
  highlight?: boolean;
}> = ({ email, isPrimary, highlight, onSetPrimary, onRemove }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment(FRAGMENT, email);
  const setPrimaryEmail = useSetAtom(setPrimaryEmailFamily(data.id));
  const removeEmail = useSetAtom(removeEmailFamily(data.id));
  const { t } = useTranslation();

  const onRemoveClick = (): void => {
    startTransition(() => {
      removeEmail().then(() => {
        // Call the onRemove callback if provided
        onRemove?.();
      });
    });
  };

  const onSetPrimaryClick = (): void => {
    startTransition(() => {
      setPrimaryEmail().then(() => {
        // Call the onSetPrimary callback if provided
        onSetPrimary?.();
      });
    });
  };

  return (
    <Form.Root>
      <Form.Field name="email">
        <Form.Label>
          {isPrimary
            ? t("frontend.user_email.primary_email")
            : t("frontend.user_email.email")}
        </Form.Label>
        <div className="flex">
          <Form.TextControl
            type="email"
            readOnly
            value={data.email}
            className={styles.userEmailField}
          />
          <DeleteButtonWithConfirmation
            disabled={isPrimary || pending}
            onClick={onRemoveClick}
          />
        </div>

        <Form.HelpMessage>
          {data.confirmedAt && !isPrimary && (
            <button
              className={styles.link}
              disabled={pending}
              onClick={onSetPrimaryClick}
            >
              {t("frontend.user_email.make_primary_button")}
            </button>
          )}
          {!data.confirmedAt && (
            <>
              <span className={styles.userEmailUnverified}>
                {t("frontend.user_email.unverified")}
              </span>{" "}
              |{" "}
              <Link kind="button" route={{ type: "verify-email", id: data.id }}>
                {t("frontend.user_email.retry_button")}
              </Link>
            </>
          )}
        </Form.HelpMessage>
      </Form.Field>
    </Form.Root>
  );
};

export default UserEmail;
