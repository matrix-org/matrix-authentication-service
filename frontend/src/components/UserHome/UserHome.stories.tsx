// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Meta, StoryObj } from "@storybook/react";
import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

import { appConfigAtom, locationAtom } from "../../Router";
import { makeFragmentData } from "../../gql";
import { FRAGMENT as EMAIL_FRAGMENT } from "../UserEmail";

import UserHome, { FRAGMENT } from "./UserHome";

type Props = {
  primaryEmail: string | null;
  unverifiedEmails: number;
  confirmedEmails: number;
  oauth2Sessions: number;
  browserSessions: number;
  compatSessions: number;
};

const WithHomePage: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: "/" }],
  ]);
  return <>{children}</>;
};

const Template: React.FC<Props> = ({
  primaryEmail: email,
  unverifiedEmails,
  confirmedEmails,
  oauth2Sessions,
  browserSessions,
  compatSessions,
}) => {
  let primaryEmail = null;
  if (email) {
    primaryEmail = {
      id: "email:123",
      ...makeFragmentData(
        {
          id: "email:123",
          email,
          confirmedAt: new Date(),
        },
        EMAIL_FRAGMENT,
      ),
    };
  }
  const data = makeFragmentData(
    {
      id: "user:123",
      primaryEmail,
      unverifiedEmails: {
        totalCount: unverifiedEmails,
      },
      confirmedEmails: {
        totalCount: confirmedEmails,
      },
      oauth2Sessions: {
        totalCount: oauth2Sessions,
      },
      browserSessions: {
        totalCount: browserSessions,
      },
      compatSessions: {
        totalCount: compatSessions,
      },
    },
    FRAGMENT,
  );
  return (
    <Provider>
      <WithHomePage>
        <UserHome user={data} />
      </WithHomePage>
    </Provider>
  );
};

const meta = {
  title: "Pages/User Home",
  component: Template,
  tags: ["autodocs"],
} satisfies Meta<typeof Template>;

export default meta;
type Story = StoryObj<typeof Template>;

export const Basic: Story = {
  args: {
    primaryEmail: "hello@example.com",
    unverifiedEmails: 0,
    confirmedEmails: 5,
    oauth2Sessions: 3,
    compatSessions: 1,
    browserSessions: 2,
  },
};

export const Empty: Story = {
  args: {
    primaryEmail: "hello@example.com",
    unverifiedEmails: 0,
    confirmedEmails: 1,
    oauth2Sessions: 0,
    compatSessions: 0,
    browserSessions: 0,
  },
};

export const UnverifiedEmails: Story = {
  args: {
    primaryEmail: "hello@example.com",
    unverifiedEmails: 1,
    confirmedEmails: 1,
    oauth2Sessions: 0,
    compatSessions: 0,
    browserSessions: 0,
  },
};
