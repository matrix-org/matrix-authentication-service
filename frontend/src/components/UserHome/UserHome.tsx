// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Alert, Body } from "@vector-im/compound-web";
import { useState } from "react";

import { Link } from "../../Router";
import { FragmentType, graphql, useFragment } from "../../gql";
import UserEmail from "../UserEmail";

const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserHome_user on User {
    id

    primaryEmail {
      id
      ...UserEmail_email
    }

    confirmedEmails: emails(first: 0, state: CONFIRMED) {
      totalCount
    }

    unverifiedEmails: emails(first: 0, state: PENDING) {
      totalCount
    }

    browserSessions(first: 0, state: ACTIVE) {
      totalCount
    }

    oauth2Sessions(first: 0, state: ACTIVE) {
      totalCount
    }

    compatSessions(first: 0, state: ACTIVE) {
      totalCount
    }
  }
`);

const UserHome: React.FC<{
  user: FragmentType<typeof FRAGMENT>;
}> = ({ user }) => {
  const data = useFragment(FRAGMENT, user);
  const [dismiss, setDismiss] = useState(false);

  const doDismiss = (): void => {
    setDismiss(true);
  };

  return (
    <>
      {data.unverifiedEmails.totalCount > 0 && !dismiss && (
        <Alert type="critical" title="Unverified email" onClose={doDismiss}>
          You have {data.unverifiedEmails.totalCount} unverified email
          address(es). <Link route={{ type: "email-list" }}>Check</Link>
        </Alert>
      )}

      {data.primaryEmail ? (
        <UserEmail email={data.primaryEmail} isPrimary />
      ) : (
        <Alert type="critical" title="No primary email adress" />
      )}

      {data.confirmedEmails.totalCount > 1 && (
        <Body>
          {data.confirmedEmails.totalCount} additional emails.{" "}
          <Link route={{ type: "email-list" }}>View all</Link>
        </Body>
      )}

      <Body>
        {data.browserSessions.totalCount} active browser session(s).{" "}
        <Link route={{ type: "browser-session-list" }}>View all</Link>
      </Body>
      <Body>
        {data.oauth2Sessions.totalCount} active OAuth2 session(s).{" "}
        <Link route={{ type: "oauth2-session-list" }}>View all</Link>
      </Body>
      <Body>
        {data.compatSessions.totalCount} active compatibility layer session(s).{" "}
        <Link route={{ type: "compat-session-list" }}>View all</Link>
      </Body>
    </>
  );
};

export default UserHome;
