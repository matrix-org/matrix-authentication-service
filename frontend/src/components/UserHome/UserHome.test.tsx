// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { create } from "react-test-renderer";
import { describe, expect, it } from "vitest";

import { makeFragmentData } from "../../gql";
import { FRAGMENT as EMAIL_FRAGMENT } from "../UserEmail";

import UserHome, { FRAGMENT } from "./";

describe("UserHome", () => {
  it("render an simple <UserHome />", () => {
    const primaryEmail = makeFragmentData(
      {
        id: "email:123",
        email: "hello@example.com",
        confirmedAt: new Date(),
      },
      EMAIL_FRAGMENT,
    );

    const user = makeFragmentData(
      {
        id: "user:123",
        primaryEmail: {
          id: "email:123",
          ...primaryEmail,
        },
        compatSessions: {
          totalCount: 0,
        },
        browserSessions: {
          totalCount: 0,
        },
        oauth2Sessions: {
          totalCount: 0,
        },
        unverifiedEmails: {
          totalCount: 0,
        },
        confirmedEmails: {
          totalCount: 1,
        },
      },
      FRAGMENT,
    );
    const component = create(<UserHome user={user} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render a <UserHome /> with additional emails", () => {
    const primaryEmail = makeFragmentData(
      {
        id: "email:123",
        email: "hello@example.com",
        confirmedAt: new Date(),
      },
      EMAIL_FRAGMENT,
    );

    const user = makeFragmentData(
      {
        id: "user:123",
        primaryEmail: {
          id: "email:123",
          ...primaryEmail,
        },
        compatSessions: {
          totalCount: 0,
        },
        browserSessions: {
          totalCount: 0,
        },
        oauth2Sessions: {
          totalCount: 0,
        },
        unverifiedEmails: {
          totalCount: 0,
        },
        confirmedEmails: {
          totalCount: 4,
        },
      },
      FRAGMENT,
    );
    const component = create(<UserHome user={user} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render a <UserHome /> without primary email", () => {
    const user = makeFragmentData(
      {
        id: "user:123",
        primaryEmail: null,
        compatSessions: {
          totalCount: 0,
        },
        browserSessions: {
          totalCount: 0,
        },
        oauth2Sessions: {
          totalCount: 0,
        },
        unverifiedEmails: {
          totalCount: 0,
        },
        confirmedEmails: {
          totalCount: 0,
        },
      },
      FRAGMENT,
    );
    const component = create(<UserHome user={user} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render a <UserHome /> with an unverified email", () => {
    const primaryEmail = makeFragmentData(
      {
        id: "email:123",
        email: "hello@example.com",
        confirmedAt: new Date(),
      },
      EMAIL_FRAGMENT,
    );

    const user = makeFragmentData(
      {
        id: "user:123",
        primaryEmail: {
          id: "email:123",
          ...primaryEmail,
        },
        compatSessions: {
          totalCount: 0,
        },
        browserSessions: {
          totalCount: 0,
        },
        oauth2Sessions: {
          totalCount: 0,
        },
        unverifiedEmails: {
          totalCount: 1,
        },
        confirmedEmails: {
          totalCount: 1,
        },
      },
      FRAGMENT,
    );
    const component = create(<UserHome user={user} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
