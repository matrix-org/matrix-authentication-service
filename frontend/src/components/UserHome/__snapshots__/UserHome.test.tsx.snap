// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserHome > render a <UserHome /> with additional emails 1`] = `
<section
  className="m-4 flex flex-col gap-2"
>
  <h3
    className="_font-heading-md-semibold_1g2sj_129"
  >
    Where you're signed in
  </h3>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Browser
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        New apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/oauth2-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Regular apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/compat-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
</section>
`;

exports[`UserHome > render a <UserHome /> with an unverified email 1`] = `
[
  <div
    className="_alert_fxw8y_19"
    data-type="critical"
  >
    <svg
      aria-hidden={true}
      className="_icon_fxw8y_52"
      fill="currentColor"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 17a.97.97 0 0 0 .713-.288A.968.968 0 0 0 13 16a.968.968 0 0 0-.287-.713A.968.968 0 0 0 12 15a.968.968 0 0 0-.713.287A.968.968 0 0 0 11 16c0 .283.096.52.287.712.192.192.43.288.713.288Zm0-4c.283 0 .52-.096.713-.287A.968.968 0 0 0 13 12V8a.967.967 0 0 0-.287-.713A.968.968 0 0 0 12 7a.968.968 0 0 0-.713.287A.967.967 0 0 0 11 8v4c0 .283.096.52.287.713.192.191.43.287.713.287Zm0 9a9.738 9.738 0 0 1-3.9-.788 10.099 10.099 0 0 1-3.175-2.137c-.9-.9-1.612-1.958-2.137-3.175A9.738 9.738 0 0 1 2 12a9.74 9.74 0 0 1 .788-3.9 10.099 10.099 0 0 1 2.137-3.175c.9-.9 1.958-1.612 3.175-2.137A9.738 9.738 0 0 1 12 2a9.74 9.74 0 0 1 3.9.788 10.098 10.098 0 0 1 3.175 2.137c.9.9 1.613 1.958 2.137 3.175A9.738 9.738 0 0 1 22 12a9.738 9.738 0 0 1-.788 3.9 10.098 10.098 0 0 1-2.137 3.175c-.9.9-1.958 1.613-3.175 2.137A9.738 9.738 0 0 1 12 22Z"
        fill="currentColor"
      />
    </svg>
    <div
      className="_content_fxw8y_44"
    >
      <p
        className="_title_fxw8y_48"
      >
        Unverified email
      </p>
      <p>
        You have 
        1
         unverified email address(es). 
        <a
          href="/profile"
          onClick={[Function]}
        >
          Check
        </a>
      </p>
    </div>
    <svg
      aria-label="Close"
      className="_close_fxw8y_68"
      fill="currentColor"
      height={16}
      onClick={[Function]}
      role="button"
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m12 13.4-4.9 4.9a.948.948 0 0 1-.7.275.948.948 0 0 1-.7-.275.948.948 0 0 1-.275-.7.95.95 0 0 1 .275-.7l4.9-4.9-4.9-4.9a.948.948 0 0 1-.275-.7.95.95 0 0 1 .275-.7.948.948 0 0 1 .7-.275.95.95 0 0 1 .7.275l4.9 4.9 4.9-4.9a.948.948 0 0 1 .7-.275.95.95 0 0 1 .7.275.948.948 0 0 1 .275.7.948.948 0 0 1-.275.7L13.4 12l4.9 4.9a.948.948 0 0 1 .275.7.948.948 0 0 1-.275.7.948.948 0 0 1-.7.275.948.948 0 0 1-.7-.275L12 13.4Z"
        fill="currentColor"
      />
    </svg>
  </div>,
  <section
    className="m-4 flex flex-col gap-2"
  >
    <h3
      className="_font-heading-md-semibold_1g2sj_129"
    >
      Where you're signed in
    </h3>
    <div
      className="_block_17898c flex flex-row justify-between align-center gap-1"
    >
      <div
        className="flex flex-col"
      >
        <h1
          className="_font-body-md-semibold_1g2sj_69"
        >
          Browser
        </h1>
        <p
          className="_font-body-md-regular_1g2sj_65"
        >
          0
           active
           
          sessions
        </p>
      </div>
      <a
        href="/sessions"
        onClick={[Function]}
      >
        View all
      </a>
    </div>
    <div
      className="_block_17898c flex flex-row justify-between align-center gap-1"
    >
      <div
        className="flex flex-col"
      >
        <h1
          className="_font-body-md-semibold_1g2sj_69"
        >
          New apps
        </h1>
        <p
          className="_font-body-md-regular_1g2sj_65"
        >
          0
           active
           
          sessions
        </p>
      </div>
      <a
        href="/oauth2-sessions"
        onClick={[Function]}
      >
        View all
      </a>
    </div>
    <div
      className="_block_17898c flex flex-row justify-between align-center gap-1"
    >
      <div
        className="flex flex-col"
      >
        <h1
          className="_font-body-md-semibold_1g2sj_69"
        >
          Regular apps
        </h1>
        <p
          className="_font-body-md-regular_1g2sj_65"
        >
          0
           active
           
          sessions
        </p>
      </div>
      <a
        href="/compat-sessions"
        onClick={[Function]}
      >
        View all
      </a>
    </div>
  </section>,
]
`;

exports[`UserHome > render a <UserHome /> without primary email 1`] = `
<section
  className="m-4 flex flex-col gap-2"
>
  <h3
    className="_font-heading-md-semibold_1g2sj_129"
  >
    Where you're signed in
  </h3>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Browser
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        New apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/oauth2-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Regular apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/compat-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
</section>
`;

exports[`UserHome > render an simple <UserHome /> 1`] = `
<section
  className="m-4 flex flex-col gap-2"
>
  <h3
    className="_font-heading-md-semibold_1g2sj_129"
  >
    Where you're signed in
  </h3>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Browser
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        New apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/oauth2-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
  <div
    className="_block_17898c flex flex-row justify-between align-center gap-1"
  >
    <div
      className="flex flex-col"
    >
      <h1
        className="_font-body-md-semibold_1g2sj_69"
      >
        Regular apps
      </h1>
      <p
        className="_font-body-md-regular_1g2sj_65"
      >
        0
         active
         
        sessions
      </p>
    </div>
    <a
      href="/compat-sessions"
      onClick={[Function]}
    >
      View all
    </a>
  </div>
</section>
`;
